### ISSUE TITLE
Neomake! scrambling make's stderr and stdout

### ISSUE THREAD

**Issue by conte91** [neomake/neomake#1884]
### Expected behavior

When running `Neomake!` with an `errorformat` containing `%D` and `%X` (to deal with file paths in recursive Makefiles), lines coming from `make`'s `stderr` and `stdout` should be considered in order, but it seems that the whole `stdout` is processed before `stderr` instead. This makes Neomake incapable of correctly recognizing the paths of files containing errors.

### Steps to reproduce
Unpack [this example program](https://github.com/neomake/neomake/files/1793411/neomake_errorfmt_test.tar.gz), open neovim in the source folder.

Configure Neomake to run `make -j1`:

```
:let g:neomake_make_maker = { 'exe': 'make', 'args': '-j1' }
:set errorformat=%f:%l:%c:\ %t%*[a-z]:\ %m,%Xmake[%*[0-9]]:\ Leaving\ directory\ %*[`']%f%*[`'],%Dmake[%*[0-9]]:\ Entering\ directory\ %*[`']%f%*[`']
```
Check that directory changes (i.e. `%D`, `%X` entries of the `errorformat`) are parsed correctly in the quickfix window, when running `:make`.

When running `:Neomake! make`, the `make[1]: Leaving directory 'subdir'` entry is output to the quickfix window before `gcc`'s error message. This confuses Neomake, which thinks `subfunc.c` is in the top-level directory.

If `make`'s `stderr` and `stdout` are merged with `:let g:neomake_make_maker = { 'exe': 'sh', 'args': '-c "make -j1 2>&1"'}`, lines are in the correct order and Neomake can process paths correctly.

#### Neomake debug information

Async support: 1
Current filetype: 
Windows: 0
[shell, shellcmdflag, shellslash]:['/bin/bash', '-c', 0]

##### Enabled makers

For the current filetype ("", used with :Neomake):
None.
NOTE: the current buffer does not have a filetype.

Default maker settings:
 - output_stream: 'both'
 - remove_invalid_entries: 0
 - buffer_output: 1

For the project (used with :Neomake!):
None.
NOTE: you can define g:neomake_enabled_makers to configure it.

###### New-style (dict, overrides old-style)

```
g:neomake: unset
b:neomake: unset
```

###### Old-style

```
g:neomake_logfile = '/tmp/neomake.log'
g:neomake_make_maker = {'args': '-j1', 'exe': 'make'}
g:neomake_place_signs = 1

```
  makeprg=make

#### :version

```
NVIM v0.2.0
Build type: RelWithDebInfo
Compilation: /usr/bin/cc -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches  -m64 -mtune=generic -Wconversion -Wp,-U_FORTIFY_SOURCE -Wp,-D_FORTIFY_SOURCE=1 -O2 -g -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/builddir/build/BUILD/neovim-0.2.0/build/config -I/builddir/build/BUILD/neovim-0.2.0/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/builddir/build/BUILD/neovim-0.2.0/build/src/nvim/auto -I/builddir/build/BUILD/neovim-0.2.0/build/include
Compiled by mockbuild

Optional features included (+) or not (-): +acl   +iconv    +jemalloc +tui      
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"
```

#### :messages

```
/home/sbaratta
/home/sbaratta/neomake_errorfmt_test
Type  :quit<Enter>  to exit Nvim

E553: No more items
"/tmp/neomake.log" 85L, 30130C
E211: File "/tmp/neomake.log" no longer available

"/tmp/neomake.log" 23L, 3348C

Copied Neomake info to clipboard ("+).
```

Log file [here](https://github.com/neomake/neomake/files/1793455/neomake.log).

**Comment by blueyed** [comment]
Is the output order from the log correct?
```
14:17:56 [D      ] [4.4:2:1] stdout: make: ['make -C subdir', ''].
14:17:56 [D      ] [4.4:2:1] stdout: make: ['make[1]: Entering directory `/home/sbaratta/neomake_errorfmt_test/subdir''', 'gcc -o subdir.o subfunc.c', ''].
14:17:56 [D +0.01] [4.4:2:1] stderr: make: ['subfunc.c: In function ‘subFunc’:', ''].
14:17:56 [D      ] [4.4:2:1] stderr: make: ['subfunc.c:5:17: error: ‘c’ undeclared (first use in this function)', '  return a + b + c;', '                 ^', 'subfunc.c:5:17: note: each undeclared identifier is reported only once for each function it appears in', ''].
14:17:56 [D      ] [4.4:2:1] stderr: make: ['make[1]: *** [subdir.o] Error 1', 'make: *** [subdir/subfunc.o] Error 2', ''].
14:17:56 [D      ] [4.4:2:1] stdout: make: ['make[1]: Leaving directory `/home/sbaratta/neomake_errorfmt_test/subdir''', ''].
```

Otherwise I've noticed that it might depend on other factors, e.g. with the first test in `tests/filter_output.vader` stderr output is handled together:
```
      > [verbose]: [1.1:3:1] Starting async job: /usr/bin/bash -c 'echo stderr >&2; echo stdout; echo stderr2 >&2'.
      > [debug  ]: [1.1:3:1] cwd: /home/daniel/src/neomake.
      > [debug  ]: [1.1:3:1] stderr: stderr_and_stdout: ['stderr', 'stderr2', ''].
      > [debug  ]: [1.1:3:1] stdout: stderr_and_stdout: ['stdout', ''].
```

This is also true for non-buffered mode (buffered mode is used with Neovim 0.2.3+ by default in Neomake, where this would be expected really).

See `:call jobstart("/usr/bin/bash -c 'echo stderr >&2; echo stdout; echo stderr2 >&2'", {'on_stdout': {->execute('echom string(a:000)', 1)}, 'on_stderr': {->execute('echom string(a:000)', 1)}})`, which results in
```
[42, ['stderr', 'stderr2', ''], 'stderr']
[42, ['stdout', ''], 'stdout']
[42, [''], 'stderr']
[42, [''], 'stdout']
```

I think using `2>&1` here is the only way to make it work.
This could be some kind of setting, so that it wraps the maker in a shell with redirection automatically, and could be used by default if `&makeprg` gets used.
What do you think?

As for Vim:
`:call job_start(["/usr/bin/bash", "-c", 'echo stderr >&2; echo stdout; echo stderr2 >&2'], {'callback': {->execute('echom string(a:000)', 1)}})` results in
```
['channel 2 open', 'stdout']
['channel 2 open', 'stderr']
['channel 2 open', 'stderr2']
```
(adding some `sleep .1` after `echo stderr >&2` makes it work as expected btw)

**Comment by blueyed** [comment]
@conte91 
Yeah.
I've started in https://github.com/neomake/neomake/pull/2054, but then deferred it for #1575 - AFAIK Windows support is currently not working and I wanted to get the test suite there to pass before doing anything fancy like this.
